File: tests/approximate.py
#!/usr/bin/env python

# 
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
# 
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the LSST License Statement and 
# the GNU General Public License along with this program.  If not, 
# see <http://www.lsstcorp.org/LegalNotices/>.
#

"""
Tests for Approximate

Run with:
   ./approximate.py
or
   python
   >>> import approximate; approximate.run()
"""
import unittest
import numpy as np
import lsst.utils.tests as utilsTests
import lsst.afw.display.ds9 as ds9
import lsst.afw.geom as afwGeom
import lsst.afw.image as afwImage
import lsst.afw.math as afwMath
import lsst.pex.exceptions as pexExcept

try:
    display
except NameError:
    display = False

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class ApproximateTestCase(unittest.TestCase):
    
    """A test case for Approximate"""
    def setUp(self):
        pass
    
    def tearDown(self):
        pass

    def makeRamp(self, binsize=1):
        #
        # make a linear ramp
        #
        ramp = afwImage.MaskedImageF(20, 40)

        x = []
        for i in range(ramp.getWidth()):
            x.append((i + 0.5)*binsize - 0.5)

        y = []
        for j in range(ramp.getHeight()):
            y.append((j + 0.5)*binsize - 0.5)

        var = 1
        rampCoeffs = (1000, 1, 1)
        for i in range(ramp.getHeight()):
            for j in range(ramp.getWidth()):
                ramp.set(j, i, (rampCoeffs[0] + rampCoeffs[1]*x[j] + rampCoeffs[2]*y[i], 0x0, var))

        return ramp, rampCoeffs, x, y

    def testLinearRamp(self):
        """Fit a ramp"""
        
        binsize = 1
        ramp, rampCoeffs, xVec, yVec = self.makeRamp(binsize)
        #
        # Add a (labelled) bad value
        #
        ramp.set(ramp.getWidth()//2, ramp.getHeight()//2, (0, 0x1, np.nan))

        if display:
            ds9.mtv(ramp, title="Input", frame=0)
        #
        # Here's the range that the approximation should be valid (and also the
        # bbox of the image returned by getImage)
        #
        bbox = afwGeom.BoxI(afwGeom.PointI(0, 0), afwGeom.PointI(binsize*ramp.getWidth()  - 1,
                                                                 binsize*ramp.getHeight() - 1))

        order = 3                       # 1 would be enough to fit the ramp
        actrl = afwMath.ApproximateControl(afwMath.ApproximateControl.CHEBYSHEV, order)
        approx = afwMath.makeApproximate(xVec, yVec, ramp, bbox, actrl)

        for i, aim in enumerate([approx.getImage(),
                                 approx.getMaskedImage().getImage(),
                                 ]):
            if i == 0 and display:
                ds9.mtv(aim, title="interpolated", frame=1)
                with ds9.Buffering():
                    for x in xVec:
                        for y in yVec:
                            ds9.dot('+', x, y, size=0.4, frame=1)
                
            for x, y in aim.getBBox().getCorners():
                self.assertEqual(aim.get(x, y), rampCoeffs[0] + rampCoeffs[1]*x + rampCoeffs[1]*y)

    def testChebyshevEqualOrder(self):
        """Check that we enforce the condition orderX == orderY"""

        utilsTests.assertRaisesLsstCpp(self, pexExcept.InvalidParameterError,
                                       lambda : 
                                       afwMath.ApproximateControl(afwMath.ApproximateControl.CHEBYSHEV, 1, 2))

    def testLinearRampAsBackground(self):
        """Fit a ramp"""

        ramp, rampCoeffs = self.makeRamp()[0:2]

        if display:
            ds9.mtv(ramp, title="Input", frame=0)
        #
        # Here's the range that the approximation should be valid (and also the
        # bbox of the image returned by getImage)
        #
        bkgd = afwMath.makeBackground(ramp, afwMath.BackgroundControl(10, 10))

        orderMax = 3                    # 1 would be enough to fit the ramp
        for order in range(orderMax + 1):
            actrl = afwMath.ApproximateControl(afwMath.ApproximateControl.CHEBYSHEV, order)

            approx = bkgd.getApproximate(actrl)
            #
            # Get the Image, the MaskedImage, and the Image with a truncated expansion
            #
            for i, aim in enumerate([approx.getImage(),
                                     approx.getMaskedImage().getImage(),
                                     approx.getImage(order - 1 if order > 1 else -1),
                                     ]):
                if display and (i == 0 and order == 1):
                    ds9.mtv(aim, title="Interpolated", frame=1)

                for x, y in aim.getBBox().getCorners():
                    val = np.mean(aim.getArray()) if order == 0 else \
                        rampCoeffs[0] + rampCoeffs[1]*x + rampCoeffs[1]*y

                    self.assertEqual(aim.get(x, y), val)
        #
        # Check that we can't "truncate" the expansion to a higher order than we requested
        #
        utilsTests.assertRaisesLsstCpp(self, pexExcept.InvalidParameterError,
                                       lambda : approx.getImage(orderMax + 1, orderMax + 1))

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""

    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(ApproximateTestCase)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)
    return unittest.TestSuite(suites)

def run(shouldExit = False):
    """Run the tests"""
    utilsTests.run(suite(), shouldExit)

if __name__ == "__main__":
    run(True)

File: python/lsst/afw/math/_approximate.cc
/*
 * LSST Data Management System
 * Copyright 2008-2016  AURA/LSST.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <https://www.lsstcorp.org/LegalNotices/>.
 */

#include <nanobind/nanobind.h>
#include <lsst/cpputils/python.h>
#include <nanobind/stl/vector.h>

#include "lsst/afw/image/MaskedImage.h"
#include "lsst/afw/math/Approximate.h"

namespace nb = nanobind;

using namespace nb::literals;

namespace lsst {
namespace afw {
namespace math {

namespace {
template <typename PixelT>
void declareApproximate(lsst::cpputils::python::WrapperCollection &wrappers, std::string const &suffix) {
    using Class = Approximate<PixelT>;

    wrappers.wrapType(
            nb::class_<Class>(wrappers.module, ("Approximate" + suffix).c_str()),
            [](auto &mod, auto &cls) {
                cls.def("getImage", &Class::getImage, "orderX"_a = -1, "orderY"_a = -1);
                cls.def("getMaskedImage", &Class::getMaskedImage, "orderX"_a = -1, "orderY"_a = -1);

                mod.def("makeApproximate",
                        (std::shared_ptr<Approximate<PixelT>>(*)(
                                std::vector<double> const &, std::vector<double> const &,
                                image::MaskedImage<PixelT> const &, lsst::geom::Box2I const &,
                                ApproximateControl const &))makeApproximate<PixelT>,
                        "x"_a, "y"_a, "im"_a, "bbox"_a, "ctrl"_a);
            });
}
void declareApproximate(lsst::cpputils::python::WrapperCollection &wrappers) {
    auto control =
            wrappers.wrapType(nb::class_<ApproximateControl>(
                                      wrappers.module, "ApproximateControl"),
                              [](auto &mod, auto &cls) {
                                  cls.def(nb::init<ApproximateControl::Style, int, int, bool>(), "style"_a,
                                          "orderX"_a, "orderY"_a = -1, "weighting"_a = true);

                                  cls.def("getStyle", &ApproximateControl::getStyle);
                                  cls.def("setStyle", &ApproximateControl::setStyle);
                                  cls.def("getOrderX", &ApproximateControl::getOrderX);
                                  cls.def("setOrderX", &ApproximateControl::setOrderX);
                                  cls.def("getOrderY", &ApproximateControl::getOrderY);
                                  cls.def("setOrderY", &ApproximateControl::setOrderY);
                                  cls.def("getWeighting", &ApproximateControl::getWeighting);
                                  cls.def("setWeighting", &ApproximateControl::setWeighting);
                              });
    wrappers.wrapType(nb::enum_<ApproximateControl::Style>(control, "Style"), [](auto &mod, auto &enm) {
        enm.value("UNKNOWN", ApproximateControl::Style::UNKNOWN);
        enm.value("CHEBYSHEV", ApproximateControl::Style::CHEBYSHEV);
        enm.value("NUM_STYLES", ApproximateControl::Style::NUM_STYLES);
        enm.export_values();
    });
    declareApproximate<float>(wrappers, "F");
}
}  // namespace
void wrapApproximate(lsst::cpputils::python::WrapperCollection &wrappers) {
    wrappers.addSignatureDependency("lsst.afw.image");
    declareApproximate(wrappers);
}
}  // namespace math
}  // namespace afw
}  // namespace lsst
